### PR TITLE
ENCODER: Don't pass libctx to OSSL_ENCODER_CTX_new_by_EVP_PKEY()

### LINKED ISSUE #13544: Does EVP_ENCODER_CTX_new_by_EVP_PKEY() need an explicit library context?
This is the current signature for that function:

``` C
OSSL_ENCODER_CTX *OSSL_ENCODER_CTX_new_by_EVP_PKEY(const EVP_PKEY *pkey,
                                                   int selection,
                                                   const char *output_type,
                                                   const char *output_struct,
                                                   OSSL_LIB_CTX *libctx,
                                                   const char *propquery);
```

However, the passed `pkey` already holds a library context, so it can just as well be extracted from there.  Furthermore, it's questionable whether the key should be passed to another library context for encoding.

### PR DESCRIPTION
**Pull request by levitte**
The passed 'pkey' already contains a library context, and the encoder
implementations should be found within the same context, so passing an
explicit library context seems unnecessary, and potentially dangerous.

Fixes #13544

### CODE REVIEW

File: crypto/encode_decode/encoder_pkey.c [hunk 1]
```diff
@@ -210,10 +210,11 @@ static void encoder_destruct_EVP_PKEY(void *arg)
 static int ossl_encoder_ctx_setup_for_EVP_PKEY(OSSL_ENCODER_CTX *ctx,
                                                const EVP_PKEY *pkey,
                                                int selection,
-                                               OSSL_LIB_CTX *libctx,
                                                const char *propquery)
 {
     struct construct_data_st *data = NULL;
+    const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
```
Comment (slontis) [inline crypto/encode_decode/encoder_pkey.c:216]: Static analyser will not like derefs on the pkey
followed by a check for pkey == NULL.

Comment (levitte) [reply crypto/encode_decode/encoder_pkey.c:216]: Ah, good point!

Comment (levitte) [reply crypto/encode_decode/encoder_pkey.c:216]: Fixed, I think

Comment (slontis) [reply crypto/encode_decode/encoder_pkey.c:216]: You fixed one of them..
I was talking about the code in ossl_encoder_ctx_setup_for_EVP_PKEY() :)

Comment (levitte) [reply crypto/encode_decode/encoder_pkey.c]: Damn, did I miss that one?  Up I did.  Fixed!

**Comment by slontis** [comment]
Tests are failing quite a bit..

**Comment by levitte** [comment]
> Tests are failing quite a bit.. Wonder if you have a null access somewhere..

I'll have a closer look tomorrow

**Review by t-j-h — APPROVED** [review]

**Comment by t8m** [comment]
The reported CI failures aren't relevant.

**Comment by levitte** [comment]
Merged

cbcbac644c4679e535948e49983d335ae46c578e ENCODER: Don't pass libctx to OSSL_ENCODER_CTX_new_by_EVP_PKEY()
b03da688a223c18b5a10b5a66abe229bbb590133 Adapt everything else to the updated OSSL_ENCODER_CTX_new_by_EVP_PKEY()

**Pull request closed without merging**